nighthawk/source/common/statistic_impl.h

```diff
@@ -0,0 +1,93 @@
+#pragma once
+
+#include <memory>
+#include <vector>
+
+#include "nighthawk/hdrhistogram_c/src/hdr_histogram.h"
+
+#include "common/common/logger.h"
+
+#include "nighthawk/common/statistic.h"
+#include "nighthawk/source/common/frequency.h"
+
+namespace Nighthawk {
+
+class StatisticImpl : public Statistic, public Envoy::Logger::Loggable<Envoy::Logger::Id::main> {
+public:
+  std::string toString() const override;
+  void toProtoOutput(nighthawk::client::Output& output) override;
+};
+
+/**
+ * Simple statistic that keeps track of count/mean/var/stdev with low memory
+ * requirements.
+ */
+class StreamingStatistic : public StatisticImpl {
+public:
+  StreamingStatistic();
+  void addValue(int64_t value) override;
+  uint64_t count() const override;
+  double mean() const override;
+  double variance() const override;
+  double stdev() const override;
+  std::unique_ptr<Statistic> combine(const Statistic& statistic) override;
+
+private:
+  uint64_t count_;
+  double mean_;
+  double sum_of_squares_;
+};
+
+/**
+ * InMemoryStatistic uses StreamingStatistic under the hood to compute statistics.
+ * Stores the raw latencies in-memory, which may accumulate to a lot
+ * of data(!). Not used right now, but useful for debugging purposes.
+ */
+class InMemoryStatistic : public StatisticImpl {
+public:
+  InMemoryStatistic();
+  void addValue(int64_t sample_value) override;
+  uint64_t count() const override;
+  double mean() const override;
+  double variance() const override;
+  double stdev() const override;
+  std::unique_ptr<Statistic> combine(const Statistic& statistic) override;
+
+private:
+  std::vector<int64_t> samples_;
+  std::unique_ptr<Statistic> streaming_stats_;
+};
+
+/**
+ * HdrStatistic uses HdrHistogram under the hood to compute statistics.
+ */
+class HdrStatistic : public StatisticImpl {
+public:
+  HdrStatistic();
+  virtual ~HdrStatistic() override;
+  void addValue(int64_t sample_value) override;
+  uint64_t count() const override;
+  double mean() const override;
+  double variance() const override;
+  double stdev() const override;
+
+  std::unique_ptr<Statistic> combine(const Statistic& statistic) override;
+
+  /**
+   * Gets a HdrStatistic instance with corrections for coordinated omission applied.
+   * This should be done only once.
```

Comment:
Why only once? This seems a fragile interface.

Reply:
You would be correcting values that are already corrected, which would be no good according to hdr_histogram docs. We could guard against that.. But it's probably not relevant, because I will remove this for now per your comment below.